File: 2025/day10.cpp
// https://adventofcode.com/2025/day/10

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <iostream>
#include <print>
#include <ranges>
#include <tuple>

using Lights = aoc::static_vector<u8, 10>;
using Button = aoc::static_vector<usize, 9>;
using Buttons = aoc::static_vector<Button, 13>;
using Joltage = aoc::static_vector<u16, 10>;
using Input = Vec<std::tuple<Lights, Buttons, Joltage>>;

namespace stdv = std::views;
namespace stdr = std::ranges;

fn parse(String const& filename) -> Input {
  using tuple = Input::value_type;
  return //
      aoc::views::read_lines(filename) |
      stdv::transform([](str line) {
        let[lights, rest] = aoc::split_once(line, "] (");
        let[buttons, joltage] = aoc::split_once(rest, ") {");
        return tuple{
            lights.substr(1) | stdv::transform([](char c) {
              return static_cast<u8>(c == '#');
            }) | aoc::ranges::to<Lights>(),
            aoc::split(buttons, ") (") | stdv::transform([](str button) {
              return aoc::split<Button>(button, ',');
            }) | aoc::ranges::to<Buttons>(),
            aoc::split<Joltage>(joltage.substr(0, joltage.size() - 1), ',')};
      }) |
      aoc::ranges::to<Input>();
}

struct DijkstraLights {
  Lights lights;
  u32 decimal;

  constexpr DijkstraLights(Lights lights = {})
      : lights(std::move(lights)),
        decimal(aoc::binary_to_number(this->lights)) {}

  fn operator==(DijkstraLights const& other) const {
    return (decimal == other.decimal);
  }
  fn operator<=>(DijkstraLights const& other) const {
    return (decimal <=> other.decimal);
  }
};

fn solve_case1(Input const& input) -> u32 {
  return aoc::ranges::accumulate(
      input | stdv::transform([](let& tuple) {
        let & [ lights, buttons, _ ] = tuple;
        let start = DijkstraLights{lights |
                                   aoc::views::transform_to_value(0) |
                                   aoc::ranges::to<Lights>()};
        let target = DijkstraLights{lights};
        let distances = aoc::shortest_distances_dijkstra(
            start,
            [&](DijkstraLights const& current) {
              return //
                  buttons |
                  stdv::transform([&](Button const& button) {
                    auto next_lights = current.lights;
                    for (let pos : button) {
                      next_lights[pos] = static_cast<u8>(next_lights[pos] == 0);
                    }
                    return DijkstraLights{next_lights};
                  }) |
                  aoc::dijkstra_uniform_neighbors_view();
            },
            target);
        return static_cast<u32>(distances.at(target));
      }),
      u32{});
}

struct DijkstraJoltage {
  Joltage joltage;
  aoc::point_type<u64> compressed;
  static constexpr let max_jolt = u64{300};

  constexpr DijkstraJoltage(Joltage joltage = {})
      : joltage(std::move(joltage)) {
    // We can compress the joltage based on observed input
    auto multiplier = u64{1};
    for (let jolt : this->joltage | stdv::take(5)) {
      compressed.x += jolt * multiplier;
      multiplier *= max_jolt;
    }
    multiplier = 1;
    for (let jolt : this->joltage | stdv::drop(5)) {
      compressed.y += jolt * multiplier;
      multiplier *= max_jolt;
    }
  }

  fn increment(usize pos) {
    joltage[pos] += 1;
    auto* value = compressed.begin() + static_cast<usize>(pos >= 5);
    *value += aoc::pown(max_jolt, pos % 5);
  }

  fn operator==(DijkstraJoltage const& other) const {
    return compressed == other.compressed;
  }
  fn operator<=>(DijkstraJoltage const& other) const {
    return compressed <=> other.compressed;
  }
};

fn solve_case2(Input const& input) -> u64 {
  return aoc::ranges::accumulate(
      input | stdv::transform([](let& tuple) {
        let & [ _, buttons, joltage ] = tuple;
        let start = DijkstraJoltage{joltage |
                                    aoc::views::transform_to_value(0) |
                                    aoc::ranges::to<Joltage>()};
        let target = DijkstraJoltage{joltage};
        let distances = aoc::shortest_distances_dijkstra(
            start,
            [&](DijkstraJoltage const& current) {
              return //
                  buttons |
                  stdv::transform([&](Button const& button) {
                    auto next = current;
                    for (let pos : button) {
                      next.increment(pos);
                    }
                    return next;
                  }) |
                  stdv::filter([&](let& joltage) {
                    return stdr::all_of( //
                        joltage.joltage | stdv::enumerate, [&](let& pair) {
                          let[index, jolt] = pair;
                          return jolt <= target.joltage[index];
                        });
                  }) |
                  aoc::dijkstra_uniform_neighbors_view();
            },
            target);
        return static_cast<u64>(distances.at(target));
      }),
      u64{});
}

int main() {
  std::println("Part 1");
  let example = parse("day10.example");
  AOC_EXPECT_RESULT(7, solve_case1(example));
  let input = parse("day10.input");
  AOC_EXPECT_RESULT(535, solve_case1(input));

  std::println("Part 2");
  std::cout.flush();
  AOC_EXPECT_RESULT(33, solve_case2(example));
  // Dijkstra is too slow for part 2 :(
  // AOC_EXPECT_RESULT(100011612, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2025/src/day10.rs
use aoc::dijkstra::DijkstraState;
use aoc::iter::CollectArrayVec;
use aoc::math::binary_to_number;
use std::cmp::Ordering;
use std::hash::{Hash, Hasher};

type Lights = arrayvec::ArrayVec<u8, 10>;
type Button = arrayvec::ArrayVec<usize, 9>;
const MAX_BUTTONS: usize = 13;
type Buttons = arrayvec::ArrayVec<Button, MAX_BUTTONS>;
type Joltage = arrayvec::ArrayVec<u16, 10>;
type Input = Vec<(Lights, Buttons, Joltage)>;

fn parse(filename: &str) -> Input {
    aoc::file::read_lines(filename)
        .iter()
        .map(|line| {
            let (lights, rest) = line.split_once("] (").unwrap();
            let (buttons, joltage) = rest.split_once(") {").unwrap();
            let lights = lights[1..]
                .chars()
                .map(|c| (c == '#') as u8)
                .collect::<Lights>();
            let buttons = buttons
                .split(") (")
                .map(|button| {
                    button
                        .split(',')
                        .map(|num| num.parse().unwrap())
                        .collect::<Button>()
                })
                .collect::<Buttons>();
            let joltage = joltage[..joltage.len() - 1]
                .split(',')
                .map(|num| num.parse().unwrap())
                .collect::<Joltage>();
            (lights, buttons, joltage)
        })
        .collect()
}

#[derive(Clone, Debug, Eq, PartialEq)]
struct DijkstraLights {
    lights: Lights,
    decimal: u32,
}
impl DijkstraLights {
    fn from(lights: Lights) -> Self {
        let decimal = binary_to_number(&lights);
        Self {
            lights: lights,
            decimal: decimal,
        }
    }
}
impl Ord for DijkstraLights {
    fn cmp(&self, other: &Self) -> Ordering {
        other.decimal.cmp(&self.decimal)
    }
}
impl PartialOrd for DijkstraLights {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
impl Hash for DijkstraLights {
    fn hash<H: Hasher>(&self, state: &mut H) {
        state.write_u32(self.decimal);
    }
}

fn solve_case1(input: &Input) -> u32 {
    input
        .iter()
        .map(|(lights, buttons, _)| {
            let start = DijkstraLights::from(lights.iter().map(|_| 0).collect::<Lights>());
            let end = DijkstraLights::from(lights.clone());
            let distances =
                aoc::dijkstra::shortest_distances(&start, &end, |current: &DijkstraLights| {
                    buttons
                        .iter()
                        .map(|button| {
                            let mut next = current.lights.clone();
                            for &pos in button {
                                next[pos] = (next[pos] == 0) as u8;
                            }
                            return DijkstraState {
                                data: DijkstraLights::from(next),
                                distance: 1,
                            };
                        })
                        .collect_array_vec::<MAX_BUTTONS>()
                });
            return *distances.get(&end).unwrap();
        })
        .sum::<u32>()
}

fn main() {
    println!("Part 1");
    let example = parse("day10.example");
    assert_eq!(7, solve_case1(&example));
    let input = parse("day10.input");
    assert_eq!(535, solve_case1(&input));

    // println!("Part 2");
    // assert_eq!(33, solve_case2(&example));
    // assert_eq!(100011612, solve_case2(&input));
}
